### PR TITLE
AF-1656 - Tracing for lifecycle transitions

### PR DESCRIPTION
**Pull request by toddbeckman-wf**
# Description

Pretty much every module could benefit from added tracing, especially if it comes free to consumers.

# Changes

- If a `globalTracer()` is configured for the application, create a span for each of the lifecycle transitions and expose it to consumers who may wish to apply additional tags or use it for `childOf` and `followsFrom` references to other tags.
- Pass on the `SpanContext` from the transition span to any child modules so their transitions have a `childOf` link back to this one.

# Testing

Play with the example and note the console logging.

### CODE REVIEW

**Comment by aviary-wf** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://dev.workiva.net/docs/information-security/aviary#configuring-aviary) coverage of security relevant code
---
_Questions or Comments? Reach out on HipChat: InfoSec Forum._

**Comment by georgelesica-wf** [comment]
lgtm!

File: lib/src/lifecycle_module.dart [hunk 1]
```diff
@@ -320,6 +347,9 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
           reason: 'A module can only be loaded once.');
     }
 
+    // TODO childOf or followsFrom
```
Comment (toddbeckman-wf) [inline lib/src/lifecycle_module.dart]: This TODO is probably done.

**Comment by evanweible-wf** [comment]
Code changes look good, just had a couple small nits.

File: lib/src/lifecycle_module.dart [hunk 2]
```diff
@@ -127,6 +134,35 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
     }.forEach(_logLifecycleEvents);
   }
 
+  /// If this module is in a transition state, this is the Span capturing the
+  /// transition state.
```
Comment (greglittlefield-wf) [inline lib/src/lifecycle_module.dart]: This doc comment could use some example usage.

**Comment by greglittlefield-wf** [comment]
Oh, and could you elaborate on the unload part?
> This does not add tracing for unload because those futures are way too sensitive to change.

File: test/test_tracer.dart [hunk 3]
```diff
@@ -0,0 +1,166 @@
+import 'dart:async';
+import 'package:opentracing/opentracing.dart';
+
+class TestSpan implements Span {
```
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply test/test_tracer.dart]: BasicTracer is not OSS

File: lib/src/lifecycle_module.dart [hunk 4]
```diff
@@ -464,7 +511,15 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
 
     Future pendingTransition;
     if (_transition != null && !_transition.isCompleted) {
-      pendingTransition = _transition.future;
+      pendingTransition = _transition.future
+          // Need to `catchError` before `whenComplete` to prevent errors from
+          // getting thrown here when they should be handled elsewhere
+          .catchError((_) {})
```
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: When `_transition` is first created, it has a `.catchError()` attached to handle it.

However, I want to add an additional `.whenComplete()` to this same future. If I didn't have the `.catchError`, it would simultaneously be caught by the other `.catchError()` and also throw in this code. I don't care about the error in `.catchError()` handler here because it's already handled. I just have it so I can call `.whenComplete()` without this thread throwing an uncaught exception.

File: lib/src/lifecycle_module.dart [hunk 5]
```diff
@@ -849,9 +926,15 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
...
-        childSuspendFutures.add(child.suspend());
+        childSuspendFutures.add(new Future(() async {
+          child._childOfContext = _activeSpan?.context;
+          return child.suspend().whenComplete(() {
+            child._childOfContext = null;
+          });
+        }));
       }
       await Future.wait(childSuspendFutures);
+      assert(activeSpan != null);
```
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: Leftover from testing

File: lib/src/lifecycle_module.dart [hunk 6]
```diff
@@ -127,6 +134,35 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
...
+  SpanContext _childOfContext;
+
+  /// Builds a span that conditionally applies a followsFrom reference if this module
+  /// was loaded by a parent module.
+  ///
+  /// Returns `null` if no globalTracer is configured.
+  Span _startTransitionSpan(String operationName) {
+    final tracer = globalTracer();
+    if (tracer == null) {
+      return null;
```
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: This is correct. I do want to exist out of this code if there is no global tracer.

**Comment by toddbeckman-wf** [comment]
I spent a whole day focused on just the unload logic, trying to figure out some combination of manipulating the many `Future`s that would give me a good start and end to the spans, but throws during unload/dispose would always break out onto threads they weren't supposed to, or the memory of the modules would be leaked, or the futures would not resolve in the order they should.

File: example/panel/modules/sample_tracer.dart [hunk 7]
```diff
@@ -0,0 +1,162 @@
+import 'dart:async';
+import 'package:opentracing/opentracing.dart';
+
+class SampleSpan implements Span {
```
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply example/panel/modules/sample_tracer.dart]: BasicTracer is not OSS

File: lib/src/lifecycle_module.dart (hunk 2 above)
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: I added an example but I'm not sure if it's what you meant.

File: lib/src/lifecycle_module.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/lifecycle_module.dart]: I'm still a little confused... Where is the other `catchError`?

Also, if you're using `whenComplete`, you shouldn't have to add a `catchError`, since it'll get called even if future completes with an error.

I'm just not sure why `catchError` is needed here when it wasn't before these changes.

File: lib/src/lifecycle_module.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/lifecycle_module.dart]: I don't think my [previous comment](https://github.com/Workiva/w_module/pull/114#discussion_r208409950) was fully addressed:
> Shouldn't this be null in cases where there's no globaTracer defined?
> ...
> Or, are we to assume globalTracer() is always non-null, and there's at least a NoopTracer available?
>
> So either:
>
> If the spans will sometimes be null, there should be test coverage for those cases (e.g., running tests with and without a global tracer set up)
> If the spans will never be null, we can remove all of the asserts or null-awares.

File: lib/src/lifecycle_module.dart [hunk 8]
```diff
@@ -849,7 +939,12 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
       _willSuspendController.add(this);
       List<Future<Null>> childSuspendFutures = <Future<Null>>[];
       for (var child in _childModules.toList()) {
-        childSuspendFutures.add(child.suspend());
+        childSuspendFutures.add(new Future(() async {
```
Comment (greglittlefield-wf) [inline lib/src/lifecycle_module.dart]: Sorry, forgot to comment on this one as well:

>#nit to help avoid race conditions in Dart 1, this could be new Future.sync

**Comment by greglittlefield-wf** [comment]
Hm, did you try something similar to the following changes? If so, what issues did that approach have?
```diff
   Future<Null> _unload(Future<Null> pendingTransition) async {
     try {
       if (pendingTransition != null) {
         await pendingTransition;
       }
+      _activeSpan = _startTransitionSpan('unload');
 
       final shouldUnloadResult = shouldUnload();
       if (!shouldUnloadResult.shouldUnload) {
         _state = _previousState;
         _previousState = null;
         _transition = null;
         // reject with shouldUnload messages
         throw new ModuleUnloadCanceledException(
             shouldUnloadResult.messagesAsString());
       }
       _willUnloadController.add(this);
       await Future.wait(_childModules.toList().map((child) {
         child._parentContext = _activeSpan?.context;
         return child.unload().whenComplete(() {
           child._parentContext = null;
         });
       }));
       await onUnload();
       if (_state == LifecycleState.unloading) {
         _state = LifecycleState.unloaded;
         _previousState = null;
         _transition = null;
       }

       _didUnloadController.add(this);
     } on ModuleUnloadCanceledException catch (error, _) {
       // In the event of a cancellation, rethrow the exception and allow the
       // caller (either unload() or onWillDispose()) to handle it.
       rethrow;
     } catch (error, stackTrace) {
       // In the event of a failed unload (the module threw an exception but did
       // not explicitly cancel the unload), emit the unload failure event and
       // then rethrow the exception so that the caller (either unload() or
       // onWillDispose()) can handle it.
       _didUnloadController.addError(error, stackTrace);
+
+      _activeSpan?.setTag('error', true);

       rethrow;
+    } finally {
+      _activeSpan?.finish();
     }
```

**Comment by greglittlefield-wf** [comment]
@toddbeckman-wf Might also be good to sync up once you get back on the stuff I've been working on around passing span references through Zone `zoneValues`, in case it comes in handy here.

File: lib/src/lifecycle_module.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: The other catchError is in the previous transition's `await _transition`

Yes, `whenComplete` will get called, but the Future still throws synchronously on line 514.

File: lib/src/lifecycle_module.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: This is the same testing issue as the other PR.

[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: I added tests for this.

File: test/lifecycle_module_test.dart [hunk 9]
```diff
@@ -277,19 +295,53 @@ Future<Null> executeStateTransition(
...
+        assert(globalTracer() == null);
+      });
+
+      runTests(false);
+    });
+  });
+}
+
+/// Returns the `globalTracer()` typecasted as a `TestTracer` or `null`.
+TestTracer globalTestTracer() {
```
Comment (toddbeckman-wf) [inline test/lifecycle_module_test.dart]: Not my favorite name.

File: test/lifecycle_module_test.dart [hunk 10]
```diff
@@ -1491,6 +1784,27 @@ void main() {
...
+            subs.add(globalTestTracer()
+                .onSpanFinish
+                .where((span) => span.tags['module.name'] == 'child')
+                .listen(expectAsync1(childSpanCompleter.complete)));
+
+            expect(parentModule.resume(),
+                throwsA(same(childModule.onResumeError)));
+
+            final span = await childSpanCompleter.future;
+            await new Future(() {}); // wait for parent to finish suspending
```
Comment (toddbeckman-wf) [inline test/lifecycle_module_test.dart]: TODO comment should be resuming

File: lib/src/lifecycle_module.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (dustinlessard-wf) [reply lib/src/lifecycle_module.dart]: > Or, are we to assume globalTracer() is always non-null. 

This is a safe assumption because by default, a NoopTracer is used.

File: lib/src/lifecycle_module.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/lifecycle_module.dart]: Sorry, I'm a little confused... Could you point out exactly where that is?

If there is a catchError there, can we instead not return the result of `catchError` to ensure it's handled by the other `catchError` and not swallowed here? 

```dart
pendingTransition = _transition.future;
_transition.future.catchError((_) {}).whenComplete(() {
  _activeSpan = _startTransitionSpan('suspend_module');
});
```

Or, if something else is calling `catchError` on the returned future and not `_transition.future`, we could also use the result of `whenComplete` without a `catchError`.

```dart
pendingTransition = _transition.future.whenComplete(() {
  _activeSpan = _startTransitionSpan('suspend_module');
});
```

File: lib/src/lifecycle_module.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply lib/src/lifecycle_module.dart]: Should we remove null-awares and tests for this, then?

**Comment by greglittlefield-wf** [comment]
One last comment around the `catchError`, otherwise LGTM

File: test/lifecycle_module_test.dart [hunk 11]
```diff
@@ -2330,4 +2335,30 @@ void runTests(bool runSpanTests) {
...
+        final sub = getTestTracer().onSpanFinish.listen((_) {
+          fail('No span should be created');
+        });
+        addTearDown(sub.cancel);
+
+        await module.load();
+        await module.suspend();
+        await module.resume();
+        await module.unload();
+        await new Future.delayed(new Duration(milliseconds: 10));
```
Comment (greglittlefield-wf) [inline test/lifecycle_module_test.dart]: What is this Future for? Could use a comment

File: lib/src/lifecycle_module.dart [hunk 12]
```diff
@@ -125,11 +132,60 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
...
+  Span get activeSpan => _activeSpan;
+
+  /// Set internally by the parent module if this module is called by [loadChildModule]
+  SpanContext _parentContext;
+
+  /// Builds a span that conditionally applies a followsFrom reference if this module
+  /// was loaded by a parent module.
+  ///
+  /// Returns `null` if this module does not override the [name] getter or if
+  /// if no globalTracer is configured.
```
Comment (greglittlefield-wf) [inline lib/src/lifecycle_module.dart]: Could we update this comment to say why? (tracing is disabled for unnamed modules since the default name gets mangled in dart2js)

File: test/lifecycle_module_test.dart [hunk 13]
```diff
@@ -711,6 +793,66 @@ void main() {
         await module.suspend();
       });
 
+      if (runSpanTests) {
+        test('should record a span', () async {
+          await gotoState(module, LifecycleState.loaded);
+          subs.add(getTestTracer()
+              .onSpanFinish
+              .where((span) => span.operationName == 'LifecycleModule.suspend')
+              .listen((span) {
```
Comment (greglittlefield-wf) [inline test/lifecycle_module_test.dart]: I think this is missing an `expectAsync1`

File: test/lifecycle_module_test.dart [hunk 14]
```diff
@@ -711,6 +793,66 @@ void main() {
...
+          await module.suspend();
+        });
+
+        test(
+            'if a transition is in progress should wait to start this until that finishes',
+            () async {
+          await gotoState(module, LifecycleState.suspended);
+
+          Completer<Span> suspendCompleter = new Completer(),
+              resumeCompleter = new Completer();
```
Comment (greglittlefield-wf) [inline test/lifecycle_module_test.dart]: #nit can we split this into two variables?

File: test/lifecycle_module_test.dart [hunk 15]
```diff
@@ -1264,19 +1515,40 @@ void main() {
...
       parentModule = new TestLifecycleModule(name: 'parent');
       childModule = new TestLifecycleModule(name: 'child');
+
+      if (runSpanTests) {
+        subs.add(getTestTracer()
+            .onSpanFinish
+            .where((span) =>
+                span.operationName == 'LifecycleModule.load' &&
+                span.tags['module.name'] == 'parent')
+            .listen((span) {
```
Comment (greglittlefield-wf) [inline test/lifecycle_module_test.dart]: Should this be wrapped in an `expectAsync1`?

File: test/lifecycle_module_test.dart [hunk 16]
```diff
@@ -990,6 +1062,46 @@ void runTests(bool runSpanTests) {
...
+          await module.resume();
+
+          final suspendSpan = await suspendCompleter.future;
+          final resumeSpan = await resumeCompleter.future;
+
+          final suspendEnd = suspendSpan.startTime.add(suspendSpan.duration);
+
+          // checks that suspend ended before or at the same time the resume started
+          expect(
+              suspendEnd.compareTo(resumeSpan.startTime), lessThanOrEqualTo(0));
```
[the remark this answers is not in the document]
Comment (toddbeckman-wf) [reply test/lifecycle_module_test.dart]:
```
 Expected: a value less than or equal to DateTime:<2018-09-05 08:53:15.174793>
        Actual: DateTime:<2018-09-05 08:53:15.174661>
         Which: is not a value less than or equal to DateTime:<2018-09-05 08:53:15.174793>
      NoSuchMethodError: Class 'DateTime' has no instance method '<'.
      Receiver: Instance of 'DateTime'
```

[the remark this answers is not in the document]
Comment (greglittlefield-wf) [reply test/lifecycle_module_test.dart]: Ah, my bad!

**Comment by greglittlefield-wf** [comment]
+10

File: lib/src/lifecycle_module.dart [hunk 17]
```diff
@@ -52,12 +53,18 @@ enum LifecycleState {
 /// Intended to be extended by most base module classes in order to provide a
 /// unified lifecycle API.
 abstract class LifecycleModule extends SimpleModule with Disposable {
+  static int _nextId = 0;
+
+  int _instanceId = _nextId++;
+
   List<LifecycleModule> _childModules = [];
   Logger _logger;
   String _name;
```
Comment (evanweible-wf) [inline lib/src/lifecycle_module.dart]: I think I'd prefer renaming this to `_defaultName` so that the check you do to see if the module author has overridden the name or not makes more sense:

```dart
if (name == _defaultName) { ... }
```

File: lib/src/lifecycle_module.dart [hunk 18]
```diff
@@ -475,7 +547,12 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
         .then(transition.complete)
         .catchError((error, trace) {
       transition.completeError(error, trace);
+      _activeSpan?.setTag('error', true);
```
Comment (evanweible-wf) [inline lib/src/lifecycle_module.dart]: Is there any value to attaching the stack trace to the span as well?

File: lib/src/lifecycle_module.dart [hunk 19]
```diff
@@ -826,7 +911,12 @@ abstract class LifecycleModule extends SimpleModule with Disposable {
       _willResumeController.add(this);
       List<Future<Null>> childResumeFutures = <Future<Null>>[];
       for (var child in _childModules.toList()) {
-        childResumeFutures.add(child.resume());
+        childResumeFutures.add(new Future.sync(() {
+          child._parentContext = _activeSpan?.context;
+          return child.resume().whenComplete(() {
+            child._parentContext = null;
```
Comment (evanweible-wf) [inline lib/src/lifecycle_module.dart]: I'm a little confused by this. Wouldn't the parent-child relationship be valid regardless of what state the modules are in? Why do we need to set and then unset the `_parentContext` reference here?

Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: The parent context refers to the current span. The parent's load context will be different from the parent's resume context, so if you just pass in one context and be done, you'll get every child's transition on the same parent's event.

File: lib/src/lifecycle_module.dart (hunk 18 above)
Comment (toddbeckman-wf) [reply lib/src/lifecycle_module.dart]: There might be although I'd imagine the logging infrastructure is better for the job, and we're already logging these errors.

File: test/lifecycle_module_test.dart [hunk 20]
```diff
@@ -1855,4 +2336,31 @@ void main() {
...
+        });
+        addTearDown(sub.cancel);
+
+        await module.load();
+        await module.suspend();
+        await module.resume();
+        await module.unload();
+        // Wait some time after transitions are over to be sure no spans are created
+        await new Future.delayed(new Duration(milliseconds: 10));
+      });
```
Comment (greglittlefield-wf) [inline test/lifecycle_module_test.dart]: Sorry, saw one more thing while reviewing your dependent PR...

Can we also add a test for an unnamed module that has loaded a child module, to assert all the null-checks on the `_activeSpan` are working as intended?

Comment (toddbeckman-wf) [reply test/lifecycle_module_test.dart]: Added

File: test/lifecycle_module_test.dart [hunk 21]
```diff
@@ -2309,6 +2315,148 @@ void runTests(bool runSpanTests) {
...
+
+    tearDown(() async {
+      await parentModule.dispose();
+      await childModule.tearDown();
+      subs.forEach((sub) => sub.cancel());
+      subs.clear();
+    });
+
+    if (runSpanTests) {
+      test('should record a span for child with no parent', () async {
```
Comment (greglittlefield-wf) [inline test/lifecycle_module_test.dart]: > for child with no parent

#nit "parent" is ambiguous here. What about something like:
> for child with unnamed parent, and thus no parentContext

Comment (toddbeckman-wf) [reply test/lifecycle_module_test.dart]: I can fix this in 1540 which is also in CR right after this one

**Comment by greglittlefield-wf** [comment]
+1

**Comment by greglittlefield-wf** [comment]
QA +1

@Workiva/release-management-pp
